pkgs/build-support/kernel/closure.rs

```diff
@@ -0,0 +1,184 @@
+use std::os::unix::ffi::{OsStrExt,OsStringExt};
+use std::ffi::{OsStr,OsString};
+use std::path::{Path,PathBuf};
+
+use std::env;
+use std::fs;
+use std::process::{Command,exit};
+use std::io::Write;
+use std::io::ErrorKind;
+
+use std::collections::HashSet;
+use std::collections::VecDeque;
+use std::hash::Hash;
+
+fn main() {
+    let args : Vec<OsString> = env::args_os().collect();
+    let allow_missing = env::var_os("allowMissing").is_some();
+    let firmware = env::var_os("firmware").unwrap();
+    let data = ModulesData {
+        kernel: Path::new(&args[1]).to_path_buf(),
+        out_dir: Path::new(&args[2]).to_path_buf(),
+        firmware_dir: Path::new(&firmware).to_path_buf(),
+        allow_missing: allow_missing,
+    };
+    let roots = (args[3..]).iter().cloned().filter_map(|root| {
+        let res = data.mod_resolve(&root);
+        if res.is_none(){
+            data.mod_missing(&root);
+        }
+        res
+    }).collect();
+
+    let cl = closure( |module| data.mod_dep(&module), &roots );
+
+    for module in cl {
+        data.copy_mod(&module);
+        data.copy_firmware(&module);
+    }
+}
+
+pub enum ModPath {
+    Path(PathBuf),
+    Builtin
+}
+
+struct ModulesData {
+    kernel: PathBuf,
+    out_dir : PathBuf,
+    firmware_dir: PathBuf,
+    allow_missing : bool
+}
+
+impl ModulesData {
```

Comment:
Not sure how much you actually need the self impl here, a bunch of functions with explicit arg-passing would proably do the trick just as well.

Reply:
changed it to use explicit passing